File: payment-service/strategies/CardProcessor.js
import { PaymentProcessor } from './PaymentProcessor.js';
import { v4 as uuidv4 } from 'uuid';

export class CardProcessor extends PaymentProcessor {
    constructor() {
        super();
        this.PAYMENT_METHOD = 'CARD';
        this.PROCESSING_DELAY_MS = 4000; // Card processing is slower
    }

    async process(transaction) {
        console.log(`Processing Card payment for transaction: ${transaction.id}`);
        
        return new Promise((resolve) => {
            setTimeout(() => {
                // Simulate success/failure (85% success rate for cards)
                const isSuccess = Math.random() < 0.85;
                
                if (isSuccess) {
                    const cardTransactionId = 'CARD' + uuidv4().substring(0, 8).toUpperCase();
                    console.log(`Card payment successful for transaction: ${transaction.id} with Card ID: ${cardTransactionId}`);
                    resolve(PaymentProcessor.success(cardTransactionId, 'Card payment processed successfully'));
                } else {
                    const failureReason = this.getRandomCardFailureReason();
                    console.warn(`Card payment failed for transaction: ${transaction.id} - ${failureReason}`);
                    resolve(PaymentProcessor.failure(failureReason));
                }
            }, this.PROCESSING_DELAY_MS);
        });
    }

    getPaymentMethod() {
        return this.PAYMENT_METHOD;
    }

    canProcess(transaction) {
        return super.canProcess(transaction) && transaction.amount >= 1; // Minimum card amount
    }

    getRandomCardFailureReason() {
        const reasons = [
            'Card declined by bank',
            'Insufficient funds',
            'Card expired',
            'Invalid CVV',
            'Transaction limit exceeded',
            'Card blocked'
        ];
        return reasons[Math.floor(Math.random() * reasons.length)];
    }
}

File: backend/src/main/java/com/upi/gateway/backend/strategy/impl/CardProcessor.java
package com.upi.gateway.backend.strategy.impl;

import com.upi.gateway.backend.model.Transaction;
import com.upi.gateway.backend.strategy.PaymentProcessor;
import com.upi.gateway.backend.strategy.PaymentResult;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;

@Component
@Slf4j
public class CardProcessor implements PaymentProcessor {
    
    private static final String PAYMENT_METHOD = "CARD";
    private static final int PROCESSING_DELAY_MS = 4000; // Card processing is slower
    
    @Override
    public CompletableFuture<PaymentResult> process(Transaction transaction) {
        log.info("Processing Card payment for transaction: {}", transaction.getId());
        
        return CompletableFuture.supplyAsync(() -> {
            try {
                // Simulate card processing time
                Thread.sleep(PROCESSING_DELAY_MS);
                
                // Simulate success/failure (85% success rate for cards)
                boolean isSuccess = ThreadLocalRandom.current().nextDouble() < 0.85;
                
                if (isSuccess) {
                    String cardTransactionId = "CARD" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
                    log.info("Card payment successful for transaction: {} with Card ID: {}", 
                            transaction.getId(), cardTransactionId);
                    return PaymentResult.success(cardTransactionId, "Card payment processed successfully");
                } else {
                    String failureReason = getRandomCardFailureReason();
                    log.warn("Card payment failed for transaction: {} - {}", transaction.getId(), failureReason);
                    return PaymentResult.failure(failureReason);
                }
                
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                log.error("Card processing interrupted for transaction: {}", transaction.getId());
                return PaymentResult.failure("Processing interrupted");
            }
        });
    }
    
    @Override
    public String getPaymentMethod() {
        return PAYMENT_METHOD;
    }
    
    @Override
    public boolean canProcess(Transaction transaction) {
        return PAYMENT_METHOD.equalsIgnoreCase(transaction.getPaymentMethod()) &&
               transaction.getAmount().doubleValue() >= 1; // Minimum card amount
    }
    
    private String getRandomCardFailureReason() {
        String[] reasons = {
            "Card declined by bank",
            "Insufficient funds",
            "Card expired",
            "Invalid CVV",
            "Transaction limit exceeded",
            "Card blocked"
        };
        return reasons[ThreadLocalRandom.current().nextInt(reasons.length)];
    }
}